source/fuzz/fuzzer_util.cpp

```diff
@@ -849,17 +855,17 @@ uint32_t UpdateFunctionType(opt::IRContext* ir_context, uint32_t function_id,
     // existing one or create a new one.
     auto type_id = FindOrCreateFunctionType(
         ir_context, new_function_type_result_id, operand_ids);
+    assert(type_id != old_function_type->result_id() &&
+           "We should've handled this case above");
 
-    if (type_id != old_function_type->result_id()) {
-      function->DefInst().SetInOperand(1, {type_id});
+    function->DefInst().SetInOperand(1, {type_id});
 
-      // DefUseManager hasn't been updated yet, so if the following condition is
-      // true, then |old_function_type| will have no users when this function
-      // returns. We might as well remove it.
-      if (ir_context->get_def_use_mgr()->NumUsers(old_function_type) == 1) {
-        old_function_type->RemoveFromList();
-        delete old_function_type;
-      }
+    // DefUseManager hasn't been updated yet, so if the following condition is
+    // true, then |old_function_type| will have no users when this function
+    // returns. We might as well remove it.
+    if (ir_context->get_def_use_mgr()->NumUsers(old_function_type) == 1) {
+      old_function_type->RemoveFromList();
+      delete old_function_type;
     }
```

Comment:
This approach can potentially cause a segfault if the user is not careful about using this function (I've added a comment in the `fuzzer_util.h` to catch user's attention). 

A potential solution might be not to remove `old_function_type` instruction here but instead have a separate fuzzer pass that would go through all instructions with result ids and remove all unused ones. I'm concerned, however, that we might end up removing some instructions added in previous transformations. For example, `FuzzerPassAddEquationInstructions` adds lots of equations that might not have any users yet but can be used to infer new synonyms and replace some ids with those synonyms later. Although, this particular example can be fixed by simply checking if the instruction participates in equation facts.

@paulthomson, @andreperezmaselco, would be good to know your opinion.

Reply:
Yes I think we generally avoid removing stuff, but function types are an exception because otherwise we might end up with a LOT of function types when we make incremental changes to function signatures. 

I think there is quite a high risk of invalidating stuff whenever you modify the module, so I am OK with this.

Reply:
Ok then.